[SPARK-6612] [MLLib] [PySpark] Python KMeans parity

FlytxtRnD opened the pull request:
The following items are added to Python kmeans:
- kmeans - setEpsilon, setInitializationSteps
- KMeansModel - computeCost, k

FlytxtRnD committed:
Python Kmeans - setEpsilon, setInitializationSteps, k and computeCost added.

FlytxtRnD committed:
Merge branch 'PyKmeansAPI', remote-tracking branch 'upstream/master' into PyKmeansAPI

FlytxtRnD committed:
style corrections.

SparkQA commented:
[Test build #29793 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29793/consoleFull) for   PR 5391 at commit [`c8700e1`](https://github.com/apache/spark/commit/c8700e1446329e52c1aeb72ef171a945b7e18f83).

SparkQA commented:
[Test build #29793 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29793/consoleFull) for   PR 5391 at commit [`c8700e1`](https://github.com/apache/spark/commit/c8700e1446329e52c1aeb72ef171a945b7e18f83).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29793/
Test PASSed.

mengxr commented on python/pyspark/mllib/clustering.py:
```diff
@@ -103,10 +117,11 @@ def load(cls, sc, path):
     @classmethod
-    def train(cls, rdd, k, maxIterations=100, runs=1, initializationMode="k-means||", seed=None):
+    def train(cls, rdd, k, maxIterations=100, runs=1, initializationMode="k-means||",
+              initializationSteps=5, epsilon=1e-4, seed=None):
```
This breaks compatibility. We didn't force keyword arguments and hence we have to keep the existing ordering of arguments.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala:
```diff
@@ -308,6 +312,15 @@ private[python] class PythonMLLibAPI extends Serializable {
+   * Java stub for Python mllib KMeansModel.computeCost()
+   */
+  def computeCostKmeansModel(
+    data: JavaRDD[Vector],
```
4-space indentation

mengxr commented on python/pyspark/mllib/clustering.py:
```diff
@@ -77,6 +79,11 @@ def clusterCenters(self):
 
+    @property
+    def k(self):
+        """Get the total number of clusters."""
```
This is a `property`. So it should documented as `Total number of clusters`.

FlytxtRnD committed:
Style corrections.

FlytxtRnD committed:
Merge branch 'PyKmeansAPI', remote-tracking branch 'upstream/master' into PyKmeansAPI

SparkQA commented:
[Test build #29841 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29841/consoleFull) for   PR 5391 at commit [`0f1f738`](https://github.com/apache/spark/commit/0f1f738fa2ea3c5c53941dd607a94b203f2abc9d).

SparkQA commented:
[Test build #29841 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29841/consoleFull) for   PR 5391 at commit [`0f1f738`](https://github.com/apache/spark/commit/0f1f738fa2ea3c5c53941dd607a94b203f2abc9d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29841/
Test PASSed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala:
```diff
@@ -291,12 +291,16 @@ private[python] class PythonMLLibAPI extends Serializable {
       maxIterations: Int,
       runs: Int,
       initializationMode: String,
+      initializationSteps: Int,
```
It would be nice to keep the same ordering of arguments as in the Python API (now that the ordering has been updated).

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala:
```diff
@@ -308,6 +312,15 @@ private[python] class PythonMLLibAPI extends Serializable {
+  def computeCostKmeansModel(
+      data: JavaRDD[Vector],
+      centers: java.util.ArrayList[Vector]): Double = {
+      new KMeansModel(centers).computeCost(data)
```
style: 2-space indentation from "def" (instead of 4)

jkbradley commented on python/pyspark/mllib/clustering.py:
```diff
@@ -89,6 +96,13 @@ def predict(self, x):
                 best_distance = distance
         return best
 
+    def computeCost(self, rdd):
```
Could you please include ```computeCost``` and ```k``` in the doc test?

jkbradley commented:
@FlytxtRnD  Thanks for the previous updates!  My 3 comments are the only issues I see.

FlytxtRnD committed:
PythonMLLibAPI:
Changed ordering of arguments
Indentation corrections

clustering.py:
doc test for computeCost and k added

FlytxtRnD committed:
PythonMLLibAPI.scala:
changed ordering of arguments and Indentation corrections

clustering.py:
doc test for computeCost and k

FlytxtRnD committed:
python 3 support for computeCost

FlytxtRnD committed:
Merge remote-tracking branch 'upstream/master' into PyKmeansAPI

FlytxtRnD committed:
Changed ordering of arguments and added doc tests

SparkQA commented:
[Test build #30483 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30483/consoleFull) for   PR 5391 at commit [`44450ab`](https://github.com/apache/spark/commit/44450ab277debbf71a6c7ebd33ebb097689b437d).

SparkQA commented:
[Test build #30483 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30483/consoleFull) for   PR 5391 at commit [`44450ab`](https://github.com/apache/spark/commit/44450ab277debbf71a6c7ebd33ebb097689b437d).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30483/
Test FAILed.

jkbradley commented:
It looks like the clustering.py doc tests failed

FlytxtRnD commented:
Ok. We'll look into the issue and will update soon.

FlytxtRnD committed:
Revert "PythonMLLibAPI:"

This reverts commit 73d08aeeba0a0c4b1a22ba3165a72a44ad6bfc65.

FlytxtRnD committed:
doc test modifications

FlytxtRnD committed:
Merge remote-tracking branch 'upstream/master' into PyKmeansAPI

SparkQA commented:
[Test build #30739 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30739/consoleFull) for   PR 5391 at commit [`f8ee3f6`](https://github.com/apache/spark/commit/f8ee3f6deaed357fa3e17162ed23d37f74f99a96).

FlytxtRnD commented:
Hi @jkbradley, @mengxr,
Is it ok to close this PR and create a new one for the same ticket? We messed up with some commits in this branch.

SparkQA commented:
[Test build #30739 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30739/consoleFull) for   PR 5391 at commit [`f8ee3f6`](https://github.com/apache/spark/commit/f8ee3f6deaed357fa3e17162ed23d37f74f99a96).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30739/
Test FAILed.

jkbradley commented:
@FlytxtRnD  You could do that, or you could fix this branch however you like and do a force push to it.
